File: hackr/weekofcode/substr.cpp
#include <iostream>
#include <vector>
using namespace std;

int maxsub(string s,string t){
    int m = s.length();
    int n = t.length();
    int ansmax = 0;
    int row[m+1]={0};
    int rown[m+1] = {0};
    for(int i = 1;i<n+1;i++){
        for(int j =1;j<m+1;j++){
            if(t[i-1]==s[j-1]){
                rown[j] = row[j-1]+1;
            }else{
                rown[j] = 0;
            }
            if(rown[j]>ansmax){
                ansmax = rown[j];
            }
        }
        for(int j=1;j<m+1;j++){
            row[j]=rown[j];
        }
    }
    return ansmax;
}

int main(int argc, char const *argv[])
{
    int n,q;
    cin>>n>>q;
    string arr[n];
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    for(int i=0;i<q;i++){
        int a,b;
        cin>>a>>b;
        cout<<maxsub(arr[a],arr[b])<<endl;
    }
    return 0;
}

File: hackr/weekofcode/substr.py
def maxSub(s,t):
    m = len(s)
    n = len(t)
    ansmax = 0
    ans = [[0 for _ in range(m+1)] for j in range(n+1)]
    for i in range(1,n+1):
        for j in range(1,m+1):
            if(t[i-1]==s[j-1]):
                ans[i][j] = ans[i-1][j-1]+1
                if(ans[i][j]>ansmax):
                    ansmax = ans[i][j]
            else:
                ans[i][j] = 0
        print ans[i]
    return ansmax

n,q = map(int,raw_input().split())
arr = []

for i in range(n):
    string = raw_input().rstrip()
    arr.append(string)

for i in range(q):
    a,b = map(int,raw_input().split())
    print maxSub(arr[a],arr[b])